### ISSUE TITLE
BackgroundCopyJobCollection enumeration throws E_ACCESSDENIED

### ISSUE THREAD

**Issue by OwnageIsMagic** [dahall/Vanara#314]
```
Vanara.PInvoke.BITS System.UnauthorizedAccessException: Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
   at Vanara.PInvoke.BITS.IBackgroundCopyJob.SetNotifyInterface(IBackgroundCopyCallback pNotifyInterface)
   at Vanara.IO.BackgroundCopyJob..ctor(IBackgroundCopyJob ijob)
   at Vanara.IO.BackgroundCopyJobCollection.Enumerator.get_Current()
```

https://github.com/dahall/Vanara/blob/77f8ca176e9424d845a3abfff82d97d1ee81b7d8/BITS/BackgroundCopyJob.cs#L132

if user token is `Administrator`, but not elevated it can see `NT SYSTEM` jobs (`job.OwnerIsElevated`) but not allowed to call `SetNotifyInterface`.
https://docs.microsoft.com/en-us/windows/win32/api/bits3_0/nf-bits3_0-ibackgroundcopyjob4-getownerelevationstate#remarks
> Note that if the job was created with an elevated token, all subsequent updates to the job must be done with an elevated token.

**Comment by dahall** [comment]
How have you reproduced this? Is this a job you created using one tool while elevated and then you're using this library while not elevated to access it?

**Comment by dahall** [comment]
Second question: Since you cannot internally elevate a process, would an acceptable response be to provide a `BackgroundCopyJob` instance that is readonly and which doesn't support events?

**Comment by OwnageIsMagic** [comment]
1. some part of windows update created this
![](https://user-images.githubusercontent.com/10534634/190925771-41fcf0b2-8828-4dcd-8633-98448a2de8cd.png)
You probably can reproduce this with `psexec` and `bitsadmin`.
2. I think you should call `SetNotifyInterface` only after user actually subscribed to any event, so user can handle exception

**Comment by dahall** [comment]
I just pushed an updated using your 2nd suggestion. Please test packages available via AppVeyor (see home page for details) and let me know if this addresses your issue.

**Comment by dahall** [comment]
I went through your comments (thank you) and made some changes and asked you some questions there. Here are the questions so you don't have to wade through the code. Let me know your thoughts.
1. On the constructor: This will throw if on a Windows version before IBackgroundCopyJob4 where `OwnerIsElevated` got introduced. I agree with your point about exceptions during debug. I'll look at a way to remove and simplify. There wasn't an obvious way without a ton more code to check OS version, admin rights, admin elevation, etc. Any thoughts on doing this simply? Are you suggesting we just let the exception get thrown when subscribing to events?
2. Calling Complete is done in case Complete has not been called and the job was transferred. What scenario do you envision would be complicated by doing this during disposal?

**Comment by OwnageIsMagic** [comment]
1. Yes, just let it propagate. It have quite obvious message attached
```
Vanara.PInvoke.BITS System.UnauthorizedAccessException: Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
   at Vanara.PInvoke.BITS.IBackgroundCopyJob.SetNotifyInterface(IBackgroundCopyCallback pNotifyInterface)
```
2. `BackgroundCopyJob` is a handle to real job managed by BITS. When closing file handle I don't expect any action, same goes here. One of BITS features is job persistence beyond process lifetime, but .net guidelines insist on disposing every IDisposable.

**Comment by dahall** [comment]
I like your point on #2 of job persistence. I'll remove the auto-Complete code and document the change. 

I'll remove the code from the constructor then. I think I'll wrap the exception and put in a little more info so it is really obvious why access is denied.

Thanks for all your help.

**Comment by OwnageIsMagic** [comment]
Also Complete may throw for variety of reasons and throwing in dispose is a bad idea

**Comment by dahall** [comment]
I just committed those changes. Sorry for all the bugs. I appreciate your help in finding and squashing them.

**Comment by OwnageIsMagic** [comment]
`AutoCompleteOnSuccess` property is also not persisted. Add notice about it in documentation.

**Comment by OwnageIsMagic** [comment]
See also https://github.com/dahall/Vanara/issues/311#issuecomment-1237552348 (for https://github.com/dahall/Vanara/commit/77f8ca176e9424d845a3abfff82d97d1ee81b7d8#diff-447817cf86df18256de9c8b3e0e78afd27d6c9a878e9d8a0bd34efaa915e023fR38)
and https://github.com/dahall/Vanara/commit/74bbe0f933cd27011c328f136ae952b53e97b03b#r84756352

**Comment by dahall** [comment]
Addressed your last items in commit

**Comment by OwnageIsMagic** [comment]
hmm, I'm sorry, but there is one more case.
move [NotifyFlags = 0;](https://github.com/dahall/Vanara/blob/e9d57433b4d9b5cef76039cefd19abff8f0a60fe/BITS/BackgroundCopyJob.cs#L831) inside `if (m_notifier is not null` (it calls `m_ijob.SetNotifyFlags(value)` in setter)
And could you make [this constructor](https://github.com/dahall/Vanara/blob/e9d57433b4d9b5cef76039cefd19abff8f0a60fe/BITS/BackgroundCopyJob.cs#L990) `public`?

**Comment by OwnageIsMagic** [comment]
Seems like it used for [command line notification](https://learn.microsoft.com/en-us/windows/desktop/api/bits1_5/nf-bits1_5-ibackgroundcopyjob2-setnotifycmdline) too, so you should not override it with 0.

**Comment by dahall** [comment]
That last one opens a can of worms. Making NotifyFlags public means some complex checking to ensure that if events are set or if that command line is set, then flags should or should not be set based on a variety of combinations.

**Comment by dahall** [comment]
I ended up deprecating `NotifyProgram` in lieu of two methods that handle the flags. Exposing the flags became very messy due to the interdependencies. Please take a look and let me know your thoughts.

**Comment by OwnageIsMagic** [comment]
`SetNotifyInterface`/`SetNotifyFlags` are intrusive and machine wide, so we can assume there is only one instance of `BackgroundCopyJob`. You can just read notifyFlags in ctor and restore this state in Dispose. (Problem we are solving that user set e.g. completion only notify flag for cmdline, subscribed to Error event and exited program -- cmd should not be called for error, but still would for complete)

**Comment by dahall** [comment]
I agree. I think I've dealt with the challenges you mention, but I used `progNotify` field to grab values in ctor. I don't restore them in `Dispose` since it can throw exceptions to set those values (as we've discussed) and since it would only work if the job is not complete. `NotifyFlags` is ignored by the native interface unless `SetNotifyInterface` or `SetNotifyCmdLine` is called anyway so it doesn't matter what the persisted value is when constructed. Unless you're seeing something I'm not, I think this is ready to close. Let me know.
